elfpy/markets/hyperdrive/hyperdrive_pricing_model.py

```diff
@@ -442,3 +448,282 @@ def calc_tokens_out_given_lp_in(
+            The maximum amount of shares that can be used to open shorts.
+        """
+        t = ONE_18 - time_stretch
+        price_factor = share_price / initial_share_price
+        k = modified_yield_space_constant(price_factor, initial_share_price, share_reserves, t, bond_reserves)
+        inner_factor = initial_share_price * longs_outstanding / share_price + minimum_share_reserves**t
+        optimal_bond_reserves = (k - price_factor * inner_factor) ** (ONE_18 / t)
+
+        # The optimal bond reserves imply a maximum short of dy = y - y0.
+        return optimal_bond_reserves - bond_reserves
+
+
+class MaxLongResult(NamedTuple):
+    """Result from calculate_max_long."""
+
+    base_amount: FixedPoint
+    bond_amount: FixedPoint
+
+
+# mimic variable from solidity.
+ONE_18 = FixedPoint("1.0")
+
+
+def calculate_max_buy(
+    z: FixedPoint, y: FixedPoint, t: FixedPoint, c: FixedPoint, mu: FixedPoint
+) -> tuple[FixedPoint, FixedPoint]:
+    r"""
+    Calculates the maximum amount of bonds that can be purchased with the specified reserves.
+
+    Parameters
+    ----------
+    z : FixedPoint
+        Amount of share reserves in the pool.
+    y : FixedPoint
+        Amount of bond reserves in the pool.
+    t : FixedPoint
+        Amount of time elapsed since term start.
+    c : FixedPoint
+        Conversion rate between base and shares.
+    mu : FixedPoint
+        Interest normalization factor for shares.
+
+    Returns
+    -------
+    tuple[FixedPoint, FixedPoint]
+        The cost in shares of the maximum bond purchase and the maximum amount of bonds that can be purchased.
+    """
+    # calculate c_div_mu by directly using regular division operator
+    c_div_mu = c / mu
+    k = modified_yield_space_constant(c_div_mu, mu, z, t, y)
+    # calculate optimal_y and optimal_z using regular division and pow operator
+    optimal_y = (k / (c_div_mu + ONE_18)) ** (ONE_18 / t)
+    optimal_z = optimal_y / mu
+
+    # calculate and return the optimal trade sizes by using regular subtraction operator
+    return (optimal_z - z, y - optimal_y)
+
+
+def modified_yield_space_constant(
+    c_div_mu: FixedPoint, mu: FixedPoint, z: FixedPoint, t: FixedPoint, y: FixedPoint
+) -> FixedPoint:
+    r"""
+    Helper function to derive invariant constant C for the YieldSpace.
+
+    Parameters
+    ----------
+    c_div_mu : FixedPoint
+        Normalized price of shares in terms of base.
+    mu : FixedPoint
+        Interest normalization factor for shares.
+    z : FixedPoint
+        Amount of share reserves in the pool.
+    t : FixedPoint
+        Amount of time elapsed since term start.
+    y : FixedPoint
+        Amount of bond reserves in the pool.
+
+    Returns
+    -------
+    FixedPoint
+        The modified YieldSpace constant C.
+    """
+    # calculate and return the modified YieldSpace constant using regular arithmetic operators
+    return c_div_mu * (mu * z) ** t + y**t
+
+
+def calculate_bonds_out_given_shares_in(
```

Comment:
This should also be in yieldspace
https://github.com/delvtech/hyperdrive/blob/bfc12304a92cb7e1b4a68c5fa9a7223c9b325ca6/contracts/src/libraries/YieldSpaceMath.sol#L309

Reply:
and it is already in elfpy: https://github.com/delvtech/elf-simulations/blob/main/elfpy/markets/hyperdrive/yieldspace_pricing_model.py#L563